Feat/meta data native contract

carpawell opened the pull request:
Contract part of #3742.

cthulhu-rider commented on pkg/innerring/internal/metachain/contracts/gas/nep17.go:
```diff
@@ -0,0 +1,215 @@
+	return stackitem.NewByteArray(u.BytesBE())
+}
+
+func (c *nep17TokenNative) postTransfer(ic *interop.Context, from, to *util.Uint160, amount *big.Int,
```
unused?

carpawell replied:
rechecked and dropped some methods

cthulhu-rider commented on pkg/innerring/internal/metachain/contracts/meta/const.go:
```diff
@@ -0,0 +1,21 @@
+	putObjectEvent = "ObjectPut"
+
+	// storage prefixes
+	metaContainersPrefix     = 0x01
```
like this could be a iota enum

cthulhu-rider commented on pkg/innerring/internal/metachain/contracts/meta/method_netmap.go:
```diff
@@ -0,0 +1,120 @@
+	Nodes keys.PublicKeys
+}
+
+func (p Placement) ToStackItem() (stackitem.Item, error) {
```
used anywhere?

cthulhu-rider commented on pkg/innerring/internal/metachain/contracts/meta/method_objects.go:
```diff
@@ -0,0 +1,207 @@
+	}
+	metaInfo, ok := metaInfoSI.Value().([]stackitem.MapElement)
+	if !ok {
+		panic(fmt.Errorf("unexpected deserialized meta information value: expected %T, %T, given", metaInfo, metaInfoSI.Value()))
```
```suggestion
		panic(fmt.Errorf("unexpected deserialized meta information value: expected %T, %T given", metaInfo, metaInfoSI.Value()))
```

cthulhu-rider commented on pkg/innerring/internal/metachain/contracts/meta/method_objects.go:
```diff
@@ -0,0 +1,207 @@
+		panic("incorrect vub")
+	}
+	if vub.Int64() <= int64(ic.BlockHeight()) {
+		panic("incorrect vub: exceeded")
```
values may be helpful

cthulhu-rider reviewed (commented):
rebase is needed

roman-khimov commented on pkg/innerring/internal/metachain/contracts/contracts.go:
```diff
@@ -0,0 +1,45 @@
+			newContracts = append(newContracts, contract)
+		case *native.Std, *native.Crypto, *native.Oracle, *native.Treasury:
+		default:
+			panic(fmt.Sprintf("unexpected native contract found: %T", contract))
```
A new contract is a significant (and _very_ rare) event, this forces explicit handling of it. Should be ok.

carpawell replied:
> is this done to not miss new contract?

yes, `*native.Treasury` was even added after this feature was started to be implemented. my idea was that once there is an update in neo-go, tests will start panicking, and we will immediately find out whether we need to adopt a new contract to our side chain or not

roman-khimov commented on pkg/innerring/internal/metachain/contracts/meta/meta.go:
```diff
@@ -0,0 +1,85 @@
+	m.AddMethod(md, desc)
+
+	desc = native.NewDescriptor("unregisterMetaContainer", smartcontract.VoidType,
+		manifest.NewParameter("cID", smartcontract.Hash256Type))
```
Here as well.

roman-khimov commented on pkg/innerring/internal/metachain/contracts/meta/method_objects.go:
```diff
@@ -0,0 +1,207 @@
+		panic("container does not support chained metadata")
+	}
+
+	sigsVectorsRaw, ok := args[1].Value().([]stackitem.Item)
```
This follows current container contract, but eventually it can be removed from parameters, nodes can sign transaction directly.

roman-khimov commented on pkg/innerring/internal/metachain/contracts/meta/method_objects.go:
```diff
@@ -0,0 +1,207 @@
+		}
+	}
+
+	// required
```
Why cid is checked separately?

roman-khimov commented on pkg/innerring/internal/metachain/contracts/meta/method_objects.go:
```diff
@@ -0,0 +1,207 @@
+	if err != nil {
+		panic(err)
+	}
+
```
That's it, no processing?

carpawell replied:
do you mean the secondary index must be in the same chain? my first implementation only needs side chain to check metadata and then notify about it. all the storage nodes handle notification on their own

roman-khimov replied:
Why do you care much about key size limits in native code?

carpawell commented on pkg/innerring/internal/metachain/contracts/contracts.go:
```diff
@@ -0,0 +1,45 @@
+			panic(fmt.Sprintf("unexpected native contract found: %T", contract))
+		}
+	}
+	return append(newContracts, meta.NewMetadata(neoContract))
```
added required contracts checks to `NewCustomNatives`

carpawell commented on pkg/innerring/internal/metachain/contracts/meta/const.go:
```diff
@@ -0,0 +1,21 @@
+	MetaDataContractID   = math.MinInt32
+	MetaDataContractName = "MetaData"
+
+	// event names
```
split internal and external things. or you meant single `const` per every commented block?

carpawell commented on pkg/innerring/internal/metachain/contracts/meta/meta.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+var (
+	_ interop.Contract = (*MetaData)(nil)
```
not forget to implement every method for the contract. every native does this in neo-go's repo

carpawell commented on pkg/innerring/internal/metachain/contracts/meta/method_containers.go:
```diff
@@ -0,0 +1,57 @@
+	if !ok {
+		panic(fmt.Errorf("unexpected argument value: %T expected, %T given", cID, args[0].Value()))
+	}
+	if len(cID) != smartcontract.Hash256Len {
```
done

roman-khimov commented on pkg/innerring/internal/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,207 @@
+
+		for _, sig := range sigVectors[i] {
+			for _, node := range placement[i].Nodes {
+				if node.Verify(sig, metaHash) {
```
It's not just about sorting. Remember, currently signatures are verified in `verify()` method of Proxy contract. This scales well. Checking them during transaction execution doesn't.

roman-khimov replied:
Multiple signatures for a single signer.

roman-khimov commented on pkg/innerring/internal/metachain/contracts/meta/meta.go:
```diff
@@ -0,0 +1,85 @@
+	eDesc := native.NewEventDescriptor(putObjectEvent,
+		manifest.NewParameter("ContainerID", smartcontract.Hash256Type),
+		manifest.NewParameter("ObjectID", smartcontract.Hash256Type),
+		manifest.NewParameter("Meta", smartcontract.MapType),
```
Check other events from native contracts.

roman-khimov reviewed (commented):
Better, but still can be optimized.

carpawell commented on pkg/innerring/internal/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,332 @@
+		panic(fmt.Errorf("cannot retrieve placement vector from stack item: %w", err))
+	}
+
+	err = isSignedBySNs(ic, placement)
```
how do i know what signer to expect? it can be any node to sign, REP out of the full placement vector. i guess i still dont get you idea. is it about a really big and complex verification script?

carpawell commented on pkg/innerring/internal/metachain/meta/method_netmap.go:
```diff
@@ -0,0 +1,119 @@
+	Nodes keys.PublicKeys
+}
+
+func (p Placement) ToStackItem() (stackitem.Item, error) {
```
we are not agreed about it: https://github.com/nspcc-dev/.github/issues/29

AnnaShaleva replied:
There are methods defined on both struct and pointer of `Placement`, it spawns warnings. Define it on pointer and they'll gone.

carpawell commented on pkg/innerring/internal/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,332 @@
+	metaInfo, ok := metaInfoSI.Value().([]stackitem.MapElement)
+	if !ok {
+		panic(fmt.Errorf("unexpected deserialized meta information value: expected %T, %T given", metaInfo, metaInfoSI.Value()))
+	}
```
this adds overhead. @roman-khimov, thoughts?

carpawell commented:
Pushed smth wrong...

roman-khimov commented on pkg/innerring/internal/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,373 @@
+
+	for i, vector := range placement {
+		var foundSigs, lastFoundSig int
+		for _, sig := range sigVectors[i] {
```
There is an optimized multisig checker in NeoGo, but likely not easily applicable.

roman-khimov reviewed (commented):
Specific indexes can be tuned, but signatures seem to be ok. Let's test it.

AnnaShaleva commented on pkg/innerring/internal/metachain/gas/gas.go:
```diff
@@ -0,0 +1,142 @@
+		append(transferParams, manifest.NewParameter("data", smartcontract.AnyType))...,
+	)
+	md = native.NewMethodAndPrice(g.Transfer, 1<<17, callflag.States|callflag.AllowCall|callflag.AllowNotify)
+	md.StorageFee = 50
```
Remove it, no storage fee. This method doesn't store anything, we don't have accounts management.

AnnaShaleva commented on pkg/innerring/internal/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,377 @@
+	if err != nil {
+		panic("incorrect vub")
+	}
+	if v, current := vub.Int64(), ic.BlockHeight(); v <= int64(current) {
```
Note that `ic.BlockHeight()` returns *persisted* block height. *Persisting* block is not taken into account. You probably should adjust this check.

AnnaShaleva commented on pkg/innerring/internal/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,377 @@
+		return fmt.Errorf("not signed by %s account", acc)
+	}
+
+	return nil
```
@roman-khimov, we just need Metadata contract to be a signer of every object put transaction instead of this. This code may be moved to `verify`. Then improper transactions won't even pass verification.

Shall we?

carpawell committed:
ir/internal: add possibility to deploy custom native contract

Also, implement new MetaData contract and GAS contract implementation without
economic.

Signed-off-by: Pavel Karpy <carpawell@nspcc.ru>

carpawell commented on pkg/innerring/internal/metachain/meta/method_netmap.go:
```diff
@@ -0,0 +1,119 @@
+			if err != nil {
+				return fmt.Errorf("%d key of %d vector is not a key: %w", j, i, err)
+			}
+
```
what exact cache do you suggest? cID -> placement? caching storage via native `GetROCache`?

AnnaShaleva commented on pkg/core/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,391 @@
+		sigVectors = append(sigVectors, vector)
+	}
+
+	cnrListRaw := ic.DAO.GetStorageItem(m.ID, append([]byte{containerPlacementPrefix}, cID[:]...))
```
`cID.BytesBE()`, ditto everywhere.

carpawell commented on go.mod:
```diff
@@ -91,7 +91,9 @@ require (
 	golang.org/x/exp v0.0.0-20250911091902-df9299821621 // indirect
+	golang.org/x/mod v0.32.0 // indirect
 	golang.org/x/text v0.34.0 // indirect
+	golang.org/x/tools v0.41.0 // indirect
```
i did not do it cause i wanted to change go.mod, this is done auto when i do `go mod tidy`

```
NSPCC/git/neofs-node  feat/meta-data-native-contract ✗                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         3m ✖  
▶ go mod why -m golang.org/x/tools
# golang.org/x/tools
github.com/nspcc-dev/neofs-node/pkg/core/metachain
github.com/nspcc-dev/neo-go/pkg/core/interop
github.com/nspcc-dev/neo-go/pkg/core/interop.test
github.com/nspcc-dev/neo-go/pkg/compiler
golang.org/x/tools/go/packages

NSPCC/git/neofs-node  feat/meta-data-native-contract ✗                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      54m ✖ ⚑  
▶ go mod why -m golang.org/x/mod  
# golang.org/x/mod
github.com/nspcc-dev/neofs-node/pkg/core/metachain
github.com/nspcc-dev/neo-go/pkg/core/interop
github.com/nspcc-dev/neo-go/pkg/core/interop.test
github.com/nspcc-dev/neo-go/pkg/compiler
golang.org/x/tools/go/packages
golang.org/x/tools/internal/gocommand
golang.org/x/mod/semver
```

these are even updated every time you update them in your repo

carpawell commented on pkg/core/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,391 @@
+			sig, ok := vectorRaw[j].Value().([]byte)
+			if !ok {
+				panic(fmt.Errorf("unexpected %d signature value in %d signatures vector: %s expected, %s given", j, i, stackitem.ByteArrayT, sigsVectorsRaw[j].Type()))
+			}
```
depend on whether an incorrect signature length is a problem. if we have enough correct signatures, attaching an incorrect one is ok to me, `verifyPlacementSignatures` would put such object

carpawell commented on pkg/core/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,391 @@
+	ic.DAO.Store.Put(makeStorageKey(ic.DAO.Version.StoragePrefix, MetaDataContractID, key), value)
+}
+
+func getStorageItem(ic *interop.Context, key []byte) state.StorageItem {
+	v, err := ic.DAO.Store.Get(makeStorageKey(ic.DAO.Version.StoragePrefix, MetaDataContractID, key))
+	if err != nil {
+		return nil
+	}
+	return v
+}
```
reused

carpawell commented on pkg/core/metachain/meta/method_containers.go:
```diff
@@ -0,0 +1,50 @@
+		panic(common.ErrAlphabetWitnessFailed)
+	}
+
+	ic.DAO.DeleteStorageItem(m.ID, append([]byte{metaContainersPrefix}, cID[:]...))
```
changed, but see no differences at all. this is the same way we take an array from `cid.ID`, `oid.ID` from our SDK and reslice it everywhere in the node, there is no bytes getter, a regular Go operation known to any go dev

https://github.com/nspcc-dev/neofs-sdk-go/blob/9bd9eff5dffb487f641847ce93eec3c4f7cc3932/container/id/id.go#L23
https://github.com/nspcc-dev/neofs-sdk-go/blob/9bd9eff5dffb487f641847ce93eec3c4f7cc3932/object/id/id.go#L25

carpawell commented on pkg/core/metachain/gas/gas.go:
```diff
@@ -0,0 +1,174 @@
+
+	paymentArgs := []stackitem.Item{
+		stackitem.NewByteArray(from.BytesBE()),
+		stackitem.NewBigInteger(amount),
```
less explicit, harder to debug, but ok

carpawell commented:
I tried my best to address every thread, but there are so many of them now, and PRs are too old, and I have many versions of it (one for the tests, one draft with the whole version, and this one), but it seems like a workable contract there now.

roman-khimov commented on pkg/core/metachain/meta/method_objects.go:
```diff
@@ -0,0 +1,373 @@
+		}
+
+		key[0] = addrIndex
+		deleteStorageItem(ic, key)
```
Yes, it's _the_ synchronization layer, it decides what's right or wrong.

carpawell replied:
ok, added typed index and allowed to delete only regular objects

carpawell replied:
left https://github.com/nspcc-dev/neofs-node/issues/3976, it should be resolved somehow

roman-khimov commented on pkg/core/metachain/meta/method_netmap.go:
```diff
@@ -0,0 +1,172 @@
+// Placement is a placeholder for container's storage nodes list.
+type Placement []PlacementVector
+
+func (p Placement) ToSCParameter() (smartcontract.Parameter, error) {
```
Potentially can be moved to some separate package.

carpawell replied:
i suggest continuing to work on this feature and see closer when there are more use cases

AnnaShaleva reviewed (commented):
https://github.com/nspcc-dev/neofs-node/pull/3759/changes#r2799473732

Linter is failing, apply `make modernize` suggestions.

Otherwise can be merged.

carpawell commented:
I have added indexed signature check so that it does not check wrong keys ever. Will try to run load tests to see if the results have changed significantly.

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts.go:
```diff
@@ -0,0 +1,66 @@
+
+// NewCustomNatives returns custom list of native contracts for metadata
+// side chain. Returned contracts:
+//   - NEO
```
neofs-node.

carpawell replied:
#3979
